python/hsfs/core/arrow_flight_client.py

```diff
@@ -0,0 +1,371 @@
+        for feature in features:
+            features[feature] = list(features[feature])
+        return featuregroups, features, filters
+
+    def _collect_featuregroups_features_and_filters_rec(self, query):
+        featuregroups = {}
+        fg = query._left_feature_group
+        fg_name = f"{fg.feature_store_name.replace('_featurestore','')}.{fg.name}_{fg.version}"  # featurestore.name_version
+        featuregroups[fg._id] = fg_name
+        filters = query._filter
+
+        features = {}
+        features[fg_name] = set([feat._name for feat in query._left_features])
+
+        if fg.event_time:
+            features[fg_name].update([fg.event_time])
+        if fg.primary_key:
+            features[fg_name].update(fg.primary_key)
+        for join in query._joins:
+            join_fg = join._query._left_feature_group
+            join_fg_name = f"{join_fg.feature_store_name.replace('_featurestore','')}.{join_fg.name}_{join_fg.version}"  # featurestore.name_version
+            if len(join._on) > 0:
+                self._update_features(
+                    features, fg_name, [feat._name for feat in join._on]
+                )
+                self._update_features(
+                    features, join_fg_name, [feat._name for feat in join._on]
+                )
+            else:
+                self._update_features(
+                    features, fg_name, [feat._name for feat in join._left_on]
+                )
+                self._update_features(
+                    features, join_fg_name, [feat._name for feat in join._right_on]
+                )
+            (
+                join_featuregroups,
+                join_features,
+                join_filters,
+            ) = self._collect_featuregroups_features_and_filters_rec(join._query)
+            featuregroups.update(join_featuregroups)
+            for join_fg_name in join_features:
+                self._update_features(
+                    features, join_fg_name, join_features[join_fg_name]
+                )
+            filters = (filters & join_filters) if join_filters is not None else filters
+
+        return featuregroups, features, filters
+
+    def _filter_to_expression(self, filters, featuregroups):
+        if not filters:
+            return None
+        filter_expression = self._resolve_logic(filters, featuregroups)
+        return filter_expression
+
+    def _resolve_logic(self, logic, featuregroups):
+        filter_expression = {}
+        filter_expression["type"] = "logic"
+        filter_expression["logic_type"] = logic._type
+        if logic._left_f:
+            filter_expression["left_filter"] = self._resolve_filter(
+                logic._left_f, featuregroups
+            )
+        elif logic._left_l:
+            filter_expression["left_filter"] = self._resolve_logic(
+                logic._left_l, featuregroups
+            )
+        else:
+            filter_expression["left_filter"] = None
```

Comment:
to avoid duplication (there could be some misspells in the key value), if changed here should also be done for right filter. Furthermore, right and left filter is done the same way, maybe make a new method, to remove this duplication.
```suggestion
        if logic._left_f:
            left_filter = self._resolve_filter(
                logic._left_f, featuregroups
            )
        elif logic._left_l:
            left_filter = self._resolve_logic(
                logic._left_l, featuregroups
            )
        else:
            left_filter = None
        filter_expression["left_filter"] = left_filter
```